fastestimator/trace/io/spot_recovery.py

```diff
@@ -0,0 +1,91 @@
+# Copyright 2019 The FastEstimator Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+import os
+
+import tensorflow as tf
+import torch
+
+from fastestimator.backend.load_model import load_model
+from fastestimator.backend.save_model import save_model
+from fastestimator.trace.trace import Trace
+from fastestimator.util.data import Data
+
+
+class SpotRecovery(Trace):
+    """A trace that can backup and load your entire training system so your training can resume from the last milestone.
+        It will automatically save and load the following:
+            * Weights of all models
+            * Optimizer state of all models
+            * Training status information (e.g, current epoch index)
+
+    Args:
+        directory: Folder directory to save and load training system.
+        frequency: Milestone saving frequency in epoch(s).
+    """
+    def __init__(self, directory: str, frequency: int = 1) -> None:
+        super().__init__(mode="train")
+        self.directory = directory
+        self.frequency = frequency
+        self.model_extension = {"tf": "h5", "torch": "pt"}
+        self.optimizer_extension = {"tf": "pkl", "torch": "pt"}
+        self.system_file = "milestone.json"
+
+    def on_begin(self, data: Data) -> None:
+        if not os.path.exists(self.directory) or not os.listdir(self.directory):
+            print("FastEstimator-MileStone: Training systems will be saved in {}".format(self.directory))
```

Comment:
'Training data' instead of 'training systems' maybe?

Reply:
done